lucene/core/src/java/org/apache/lucene/util/bkd/BKDWriter.java

```diff
@@ -981,15 +988,15 @@ public long finish(IndexOutput out) throws IOException {
     assert pointCount / numLeaves <= maxPointsInLeafNode: "pointCount=" + pointCount + " numLeaves=" + numLeaves + " maxPointsInLeafNode=" + maxPointsInLeafNode;
 
     // Sort all docs once by each dimension:
-    PathSlice[] sortedPointWriters = new PathSlice[numDims];
+    PathSlice[] sortedPointWriters = new PathSlice[numDataDims];
 
     // This is only used on exception; on normal code paths we close all files we opened:
     List<Closeable> toCloseHeroically = new ArrayList<>();
 
     boolean success = false;
     try {
       //long t0 = System.nanoTime();
-      for(int dim=0;dim<numDims;dim++) {
+      for(int dim=0;dim<numDataDims;dim++) {
```

Comment:
I think we should only create sorted point writers for indexed dims. There is no point sort data dims as it only creates overhead when writing the tree.

Reply:
I made the change but then I remembered that the leaf nodes (which write all dimensions) require even the non indexed dimensions be sorted for compression (in order to compute the common prefix). I could explore delaying that sort operation only when its needed, but I think the recursion still needs sorted point writers passed through PathSlice[] so this is still necessary?